### PR TITLE
Add new args to `BatchSizeFinder`

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

Fixes https://github.com/Lightning-AI/pytorch-lightning/issues/21154
Fixes https://github.com/Lightning-AI/pytorch-lightning/issues/20447
Fixes https://github.com/Lightning-AI/pytorch-lightning/issues/20231
and
Fixes https://github.com/Lightning-AI/pytorch-lightning/issues/20614

Adds two new arguments to batch size finder: `margin` which for `mode="binsearch"` will reduce the found batch size to `int(found_batch_size * (1-margin))` to make sure there is some room for GPU allocation afterwards. The first 3 issues linked are different scenarios related to the fact that binsearch sometimes can find a batch size that just barely fits and then when training starts it crashes immediately because the found batch size does not fit in memory anymore.

The second argument added `max_val` is a simple way for users to specify a maximum batch size that can be found. One case for this arg is to make sure the algorithm does not search for a batch size larger than the dataset.

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

**Comment by Borda** [comment]
@SkafteNicki, there are about 9 failing tests...

**Comment by SkafteNicki** [comment]
putting on hold until #21187 is merged

File: src/lightning/pytorch/callbacks/batch_size_finder.py [hunk 1]
```diff
@@ -118,6 +122,8 @@ def __init__(
         init_val: int = 2,
         max_trials: int = 25,
         batch_arg_name: str = "batch_size",
+        margin: float = 0.05,
+        max_val: Optional[int] = None,
```
Comment (deependujha) [inline src/lightning/pytorch/callbacks/batch_size_finder.py]:
```suggestion
        max_val: int = 1024,
```

Let's specify a default `max_val`, so users don't end up with `batch size: ` $2^{25}$.

File: src/lightning/pytorch/tuner/batch_size_scaling.py [hunk 2]
```diff
@@ -32,6 +32,8 @@ def _scale_batch_size(
     init_val: int = 2,
     max_trials: int = 25,
     batch_arg_name: str = "batch_size",
+    margin: float = 0.05,
+    max_val: Optional[int] = None,
```
Comment (deependujha) [inline src/lightning/pytorch/tuner/batch_size_scaling.py]:
```suggestion
    max_val: int = 1024,
```

File: src/lightning/pytorch/tuner/batch_size_scaling.py [hunk 3]
```diff
@@ -58,6 +60,10 @@ def _scale_batch_size(
             - ``model.hparams``
             - ``trainer.datamodule`` (the datamodule passed to the tune method)
 
+        margin: Margin to reduce the found batch size by to provide a safety buffer. Only applied when using
+            'binsearch' mode. Should be a float between 0 and 1. Defaults to 0.05 (5% reduction).
+        max_val: Maximum batch size limit. If provided, the found batch size will not exceed this value.
```
Comment (deependujha) [inline src/lightning/pytorch/tuner/batch_size_scaling.py]:
```suggestion
        max_val: Maximum batch size limit, to prevent overly large or inefficient batch sizes.
```

File: src/lightning/pytorch/callbacks/batch_size_finder.py [hunk 4]
```diff
@@ -63,6 +63,10 @@ class BatchSizeFinder(Callback):
             - ``model.hparams``
             - ``trainer.datamodule`` (the datamodule passed to the tune method)
 
+        margin: Margin to reduce the found batch size by to provide a safety buffer. Only applied when using
+            'binsearch' mode. Should be a float between 0 and 1. Defaults to 0.05 (5% reduction).
+        max_val: Maximum batch size limit. If provided, the found batch size will not exceed this value.
```
Comment (deependujha) [inline src/lightning/pytorch/callbacks/batch_size_finder.py]:
```suggestion
        max_val: Maximum batch size limit, to prevent overly large or inefficient batch sizes.
```

File: src/lightning/pytorch/tuner/batch_size_scaling.py [hunk 5]
```diff
@@ -328,6 +352,12 @@ def _adjust_batch_size(
         pass
 
     new_size = value if value is not None else int(batch_size * factor)
+
+    # Apply max_val limit if provided
+    if max_val is not None and new_size > max_val:
+        if desc:
+            rank_zero_info(f"Batch size {new_size} exceeds max_val limit {max_val}, capping at {max_val}")
+        new_size = max_val
```
Comment (deependujha) [inline src/lightning/pytorch/tuner/batch_size_scaling.py:381]: Stop early if `max_val` is reached, no need to keep looping in `_run_binsearch_scaling` or `_run_power_scaling`.

**Comment by deependujha** [comment]
Some updates have been made to the PR: including adding a default value for `max_val` and fixing the failing tests.

Since the previous block has been resolved, it’d be great if this could be reviewed again to make sure everything’s good.

cc: @Borda @bhimrazy @SkafteNicki

**Review by rittik9 — APPROVED** [review]
